### PR TITLE
[Support] Add iterator interface to SymbolCache

### PR DESCRIPTION
**Pull request by mortbopet**
Continuing from the work in #3137, instead of introducing a new class for generating unique names, it might be better to simply have a way to initialize a namespace from a `SymbolCache`.

To do so, this PR introduces an iterator interface to the base `SymbolCache` class which iterates over pairs of `{mlir::Attribute, mlir::Operation*}`.

Given this interface, it's now possible to range-loop over any given cache and use the values for namespace initialization.

### CODE REVIEW

File: include/circt/Support/Namespace.h [hunk 1]
```diff
@@ -31,6 +32,15 @@ class Namespace {
   Namespace(const Namespace &other) = default;
   Namespace(Namespace &&other) : nextIndex(std::move(other.nextIndex)) {}
 
+  /// SymbolCache initializer; initialize from every key that is convertible to
+  /// a StringAttr in the SymbolCache.
+  Namespace(SymbolCache &symCache) {
+    for (auto &&[attr, _] : symCache) {
```
Comment (teqdruid) [inline include/circt/Support/Namespace.h:38]: nit: brackets not necessary

**Review by teqdruid — COMMENTED** [review]
This seems enormously useful. I didn't even know that the `Namespace` utility existed!

The amount and complexity of implementing iterators seems like overkill. Why not require extending classes implement a `begin()` and `end()` method which return 
`{Attribute, Operation*}` iterators? That way `SymbolCache` can just pass through to its map and `HWSymCache` can just use the llvm mapping iterator?

**Comment by mortbopet** [comment]
> Why not require extending classes implement a begin() and end() method which return {Attribute, Operation*} iterators? That way SymbolCache can just pass through to its map and HWSymCache can just use the llvm mapping iterator?

That would be the ideal, and was what I initially was going for - however, I found no simpler way of doing exactly this than the solution in the PR.

I sense that I am missing something obvious here, i.e. some form of base `llvm::forward_iterator_interface<Attribute, Operation*>` class which other iterators with identical `operator*` return type can be copy-constructed to (which would allow for `llvm::map_iterator`)... but scouring through the tree I found nothing. In general, iterator-related supporting logic is written for static polymorphism (for good reason), and not dynamic, as is required in this case.

**Comment by teqdruid** [comment]
Yeah, I just tried to what I suggested and ran into type issues. I too couldn't figure out how to do it, but I wasn't willing to spend too much time on it.

**Review by teqdruid — COMMENTED** [review]
This approach looks totally reasonable to me, but I don't know enough about how iterators work to really give this a detailed review or know if this is indeed the best way. @darthscsi (or someone more knowledgeable than I) can you give this a look?

**Review by darthscsi — APPROVED** [review]
I agree there seems like there should be a cleaner way, but I'm not seeing it.

**Pull request merged**